compiler/IREmitter/Payload.cc

```diff
@@ -985,8 +987,29 @@ llvm::Value *Payload::varGet(CompilerState &cs, cfg::LocalRef local, llvm::IRBui
     }
     if (irctx.escapedVariableIndices.contains(local)) {
         auto *cfp = getCFPForBlock(cs, builder, irctx, rubyBlockId);
-        auto [index, level] = escapedVariableIndexAndLevel(cs, local, irctx, rubyBlockId);
-        return builder.CreateCall(cs.getFunction("sorbet_readLocal"), {cfp, index, level});
+        auto info = escapedVariableInfo(cs, local, irctx, rubyBlockId);
+        auto *value = builder.CreateCall(cs.getFunction("sorbet_readLocal"), {cfp, info.index, info.level});
+        // If we ever wrote to this local, we cannot guarantee that the type has
+        // been checked prior to the access from the locals array.
+        if (info.use.used == LocalUsedHow::WrittenTo) {
+            return value;
+        }
+
+        core::ClassOrModuleRef klass;
+        typecase(
+            info.use.type, [&klass](const core::ClassType &ct) { klass = ct.symbol; },
+            [&klass](const core::AppliedType &at) { klass = at.klass; }, [](const core::TypePtr &def) {});
+        if (!klass.exists()) {
+            return value;
+        }
+
+        if (!hasOptimizedTest(klass)) {
+            return value;
+        }
+
+        Payload::assumeType(cs, builder, value, klass);
```

Comment:
It's unfortunate that you end up having to replicate the `core::TypePtr` overload of `hasOptimizedTest` here. What do you think about these two options:

* Modify `hasOptimizedTest` to return an `optional<core::ClassOrModuleRef>` to indicate that the optimized test exists
* Add an overload to `assumeType` that accepts a `core::TypePtr` argument and only emits the annotation if the type given is a `ClassType` or an `AppliedType`

My preference would be the first option, where we could also rely on `.exists()` to indicate success instead of using an `optional`.